### DISCUSSION TITLE
Cannot find module './server.node' when using ReactDOMServer.renderToStaticMarkup

### DISCUSSION THREAD

**Discussion by zoul** [vercel/next.js#42671] [Help]
Hi! I want to render a React component to static markup in an API endpoint, but I‘m getting the following error:

```
Error: Cannot find module './server.node'
Require stack:
- /project/node_modules/next/dist/compiled/react-dom/server.js
```

There’s no `server.node.js` in `node_modules/next/dist/compiled/react-dom/`, but there is one in `node_modules/react-dom`. What am I doing wrong? (Next 13.0.0 if that makes a difference.)

**Comment by icyJoseph** [comment]
Got a repo to look at? I don't want to make any assumptions about how this being done. I am also very curious :) 

One issue might be that Next.js pre-compiles dependencies, and perhaps it skips including `server.node` during this pre-compile step.

**Reply by zoul** [reply]
Thank you! Give me a moment to clean up the code a bit so that I can post it.

**Reply by zoul** [reply]
The [PR is here](https://github.com/Ohlasy/web/pull/92), the [failing function is deployed here](https://web-git-next-ohlasy.vercel.app/api/feeds/articles?plain=y) ([source here](https://github.com/Ohlasy/web/blob/next/pages/api/feeds/articles.ts)). I think the main problem is calling `ReactDOMServer.renderToStaticMarkup`. The production error is:

```
2022-11-10T06:58:03.068Z	8ba69061-1fdd-4062-b90e-b389e809ba09	ERROR	Error: Cannot find module './server.node'
Require stack:
- /var/task/node_modules/next/dist/compiled/react-dom/server.js
- /var/task/.next/server/pages/api/feeds/articles.js
- /var/task/node_modules/next/dist/server/next-server.js
- /var/task/___next_launcher.cjs
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.mod._resolveFilename (/var/task/node_modules/next/dist/build/webpack/require-hook.js:23:32)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:101:18)
    at Object.<anonymous> (/var/task/node_modules/next/dist/compiled/react-dom/server.js:3:18)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/var/task/node_modules/next/dist/compiled/react-dom/server.js',
    '/var/task/.next/server/pages/api/feeds/articles.js',
    '/var/task/node_modules/next/dist/server/next-server.js',
    '/var/task/___next_launcher.cjs'
  ]
}
RequestId: 8ba69061-1fdd-4062-b90e-b389e809ba09 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

Context: I am rewriting our newspaper’s website to Next. What I am trying to do here is to render article content from React components into a static HTML markup to use in RSS feed. It’s quite possible I’m doing something infinitely stupid, but I need the toolchain to tell me :)

Any help is appreciated, happy to add details. Thank you!

**Reply by icyJoseph** [reply]
Right, my first impression is that: Because the API routes in Vercel run in tasks, which means they do not run "with your server", rather they are serverless _**"pure functions"**_, (sort of,  the main thing is that to keep them lightweight only certain things are copied over to the task runner) then I do believe the task doesn't have `node_modules` for React. I'll take a look.

**Reply by icyJoseph** [reply]
Btw, you could get ahead of me by trying to dynamically import `react-dom/server`, using `process.cwd()`, I am not sure if it'll work but it just might! `const  ReactDOMServer  = await import(process.cwd() + "/node_modules/" + "react-dom/server")` - long shot but if you have the time to try it out, please do so.

**Reply by zoul** [reply]
Thank you very much! That results in the module not getting found, seemingly no matter how I try to tweak the path. At one moment it even says “module not found” and gives a path that works on the command line… Unfortunately I‘m a complete noob in the JavaScript ecosystem under the hood.

**Reply by icyJoseph** [reply]
Yeah, well I gave this a go, and run into issues with React state hooks not being available, and such, I also learned to load modules from CDN's within node.js and it kinda worked but looked hacky af.

> ❗️ ❗️❗️ Do not just copy paste or follow these steps blindly

I took a step back and tried it with aliases, please do try it in Vercel env:

1. Install react and react-dom with aliases:
```shell
yarn add react-ssr-dom@npm:react-dom@latest
```

`npm` also supports this via: `npm install <ALIAS_NAME>@npm:<REAL_PACKAGE>@<PACKAGE_VERSION>` but I haven't used this solution

2. Now you can work with  `react-dom-ssr` which is just an alias to the real thing
3. Create and `api/` with `.tsx` extension 
```
// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import type { NextApiRequest, NextApiResponse } from "next";
import { Article } from "../../components/Article";

// @ts-expect-error
import { renderToStaticMarkup } from "react-ssr-dom/server";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const html = renderToStaticMarkup(
    <Article title="title" description="description" />
  );

  return res.status(200).send(html);
}
```
4. Call it!
![Screenshot 2022-11-10 at 10 34 45](https://user-images.githubusercontent.com/21013447/201054336-0d6201a3-f040-469b-b3da-ff3812085b08.png)


Where the JSX is:

```jsx
import { useEffect, useState } from "react";

export const Article = (props: { title: string; description: string }) => {
  const [state, setState] = useState(null);
  useEffect(() => {
    // noop
  }, []);

  return (
    <article>
      <header>{props.title}</header>
      <main>{props.description}</main>
    </article>
  );
};
```

That was interesting, but I am afraid it might fail when deployed anyway :( please do give it a go.

**Reply by zoul** [reply]
Wow, much appreciated! 🙏  At this moment, though, I am thinking of going back to the blackboard. What I am trying to accomplish here is to create a static RSS feed. It can be generated statically just as any other page in the website tree, but Next doesn’t seem to support that yet? (create a file with different content type), so I’m using a cached API route as a workaround. So given the complexity of the suggested API route solution, isn’t there a way to make Next output a file with a different content type during the static generation phase? Or, if that isn’t feasible, I would probably try to use Markdoc to generate the HTML in the API route without depending on React; that could also make the problem disappear?
